ejb3/src/main/java/org/jboss/as/ejb3/deployment/processors/EjbJndiBindingsDeploymentUnitProcessor.java

```diff
@@ -79,11 +81,18 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
 
         final EEModuleDescription eeModuleDescription = deploymentUnit.getAttachment(Attachments.EE_MODULE_DESCRIPTION);
         final Collection<ComponentDescription> componentDescriptions = eeModuleDescription.getComponentDescriptions();
+        Map<String, EJBComponentDescription> ejbClassTypeDefined = new HashMap<>();
         if (componentDescriptions != null) {
             for (ComponentDescription componentDescription : componentDescriptions) {
                 // process only EJB session beans
                 if (componentDescription instanceof SessionBeanComponentDescription) {
-                    this.setupJNDIBindings((EJBComponentDescription) componentDescription, deploymentUnit);
+                    EJBComponentDescription ejbComponentDescription = (EJBComponentDescription) componentDescription;
+                    String ejbClassName = ejbComponentDescription.getEJBClassName();
+                    if ((ejbClassTypeDefined.putIfAbsent(ejbClassName,ejbComponentDescription) == null || !((ejbClassTypeDefined.get(ejbClassName).isSingleton() && (ejbComponentDescription.isStateful() || ejbComponentDescription.isStateless())) || (ejbClassTypeDefined.get(ejbClassName).isStateful() && (ejbComponentDescription.isSingleton() || ejbComponentDescription.isStateless())) || (ejbClassTypeDefined.get(ejbClassName).isStateless() && (ejbComponentDescription.isSingleton() || ejbComponentDescription.isStateful()))))) {
```

Comment:
I don't understand the need for these other conditions.  An EJB class name should only ever have 1 ejb component description (regardless of it's type).  Consequently, a simple putIfAbsent should be all that is needed to ensure there are not multiple type declarations.
To me, these other checks look redundant.  Am I missing something?

Reply:
Hello Paul. I am still doing some checks, but maybe there could exist multiple sessions (of same type) with different names from the same class  that should be allowed to be created.

Reply:
> maybe there could exist multiple session beans (of same type) with different names from the same class

I don't think this is possible, but perhaps I have not fully understood your meaning.  Can you provide an example?
AIUI, an EJB class has a single component description.  That description might have multiple views (e.g. \@Local vs \@Remote), but there is still only a single EJB component description which is of a single session type (\@Stateless vs \@Stateful vs \@Singleton).

Reply:
For example ...

<enterprise-beans>
<session>
<ejb-name>Name1</ejb-name>
<ejb-class>EjbClass1</ejb-class>
<session-type>Stateless</session-type>
</session>
<session>
<ejb-name>Name2</ejb-name>
<ejb-class>EjbClass1</ejb-class>
<session-type>Stateless</session-type>
</session>
</enterprise-beans>

Reply:
Sorry about the format ...

Reply:
I don't see anything in the EJB specification that suggests that a given module can contain multiple beans with the same `ejb-class` (using distinct `ejb-name`s).  On the contrary, this is explicitly disallowed via EJB annotations, evident by the fact that:

* the name property of the `@Stateless`, `@Stateful`, and `@Singleton` annotations is a String, not a String[], e.g. `@Stateless(name = { "name1", "name2" })`
* the `@Stateless`, `@Stateful`, and `@Singleton` annotations are not `@Repeatable`, which would otherwise allow the user to define multiple names for the same bean class, e.g.`@Stateless(name = "name1") @Stateless(name = "name2")`

What possible use case would require multiple beans sharing the same implementation class within the same module?

Reply:
Ejbs with different names could be enriched with different interceptors ...

Reply:
OK - I see how that could be useful.  We can still allow this and while simplifying the expression above.
e.g.
```java
        Map<String, SessionBeanComponentDescription> sessionBeanComponentDescriptions = new HashMap<>();
        if (componentDescriptions != null) {
            for (ComponentDescription componentDescription : componentDescriptions) {
                // process only EJB session beans
                if (componentDescription instanceof SessionBeanComponentDescription) {
                    SessionBeanComponentDescription sessionBeanComponentDescription = (SessionBeanComponentDescription) componentDescription;
                    String ejbClassName = sessionBeanComponentDescription.getEJBClassName();
                    SessionBeanComponentDescription existingDescription = sessionBeanComponentDescription.putIfAbsent(ejbClassName, sessionBeanComponentDescription);
                    if ((existingDescription == null) || existingDescription.getSessionBeanType() == sessionBeanComponentDescription.getSessionBeanType()) {
                        this.setupJNDIBindings(ejbComponentDescription, deploymentUnit);
                    } else {
                        EjbLogger.DEPLOYMENT_LOGGER.typeSpecViolation(ejbClassName);
                    }
                }
            }
        }
```